### DISCUSSION TITLE
Generic Narrowing Question

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#7124] [Q&A]
Had a quick question about narrowing with generics:

**Doesn't work:**
```python
from typing import TypeVar, Iterable

T = TypeVar("T")
Z = TypeVar("Z")


def foo(var: T |Iterable[Z]) -> list[Z | T]:
    return (
        list(var)
        if isinstance(var, Iterable) and not isinstance(var, str)
        else [var]
    )
```

I assumed that that `isinstance(var, Iterable)` would consistently narrow `var` into `T` however, to get this to work properly I have to add `str` into the unioned return type.

**Works:**
```python
from typing import TypeVar, Iterable

T = TypeVar("T")
Z = TypeVar("Z")


def foo(var: T | Iterable[Z]) -> list[Z | T | str]:
    return (
        list(var)
        if isinstance(var, Iterable) and not isinstance(var, str)
        else [var]
     )
```

It seems `not isinstance(var, str)` expands `T` into `T | str`. Even if I bind `T` to possibly being `str`. Would be curious to understand why this is the case.

This also doesn't work, which surprised me:

```python
T = TypeVar("T", str, int, float)
Z = TypeVar("Z")


def foo(var: T | Iterable[Z]) -> list[Z | T]:
    return (
        list(var) if isinstance(var, Iterable) and not isinstance(var, str) else [var] # We know `var` can be a string, why does it get expanded?
    )
```

**Comment by erictraut** [comment]
It's easier to see what's going on here by expanding the ternary operator into `if` and `else` statements.

```python
def foo(var: T | Iterable[Z]) -> list[Z | T]:
    if isinstance(var, Iterable):
        reveal_type(var) # Iterable[Unknown]* | Iterable[Z]
        if not isinstance(var, str):
            reveal_type(var) # Iterable[Unknown]* | Iterable[Z]
            return list(var)
        else:
            reveal_type(var) # str* | str
    else:
        reveal_type(var) # object*

    reveal_type(var)  # str* | str | object*
    return [var]
```

Where you see a `*` next to a type, that represents a [conditional type](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables). In this context, you can think of it as an intersection between the displayed type and `T`. In other words, `object*` is type compatible with both `object` and `T`.

Hopefully that makes this clearer for you. Feel free to post more questions if this doesn't make sense.
